### PR TITLE
implement queue debug command

### PR DESCRIPTION
**Pull request by ggilmore**
While debugging an issue, @stefanhengl and I both found it quite difficult to understand what the contents of the zoekt index queue were, and where in the queue a given repository might be. 

This PR implements two things:

- A new `/debug/queue` route that exposes route that retrieves every queue entry in sorted order
- A new `zoekt-sourcegraph-indexserver debug queue` command that uses the above to produce the following output:
   - [asciicast](https://asciinema.org/a/t1cGM3dF3USLcSa44RxZecwE1)
   - (note the brief change in the queue ordering after I make a new commit)

```
❯ zoekt-sourcegraph-indexserver debug queue --help                                                                                                   12:19:30
USAGE
  queue [flags]

FLAGS
  -header=true         print column headers
  -hostname localhost  the hostname of the zoekt-sourcegraph-indexserver instance to connect to
  -port 6072           the port of the zoekt-sourcegraph-indexserver instance to connect to
  -timeout 10s         max timeout for establishing a connection to the zoekt-sourcegraph-indexserver instance
```

### CODE REVIEW

**Comment by keegancsmith** [comment]
:+1: yes please. You can improve the debug page which currently is too simple. I would take inspiration from the debug page for the repo-updater queue: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@61c5905e86247a3b9f710ee777eef25a07fc27d0/-/blob/cmd/repo-updater/shared/main.go?L332&subtree=true

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: I should add a [httptest-based](https://pkg.go.dev/net/http/httptest) test to test this flow.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: These fields needed to be exported in order for gob to be able to properly read them.

Comment (ggilmore) [reply cmd/zoekt-sourcegraph-indexserver/queue.go]: https://pkg.go.dev/encoding/gob

> ...Structs encode and decode only exported fields...

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: While writing this abstraction, it occurred to me that there is nothing really tying this implementation to only queue items. I think we could make a generic `gobstream` utility package or something similar to provide a `encoder / decoder` abstraction around passing around an unbounded stream of generic items. Thoughts?

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: I noticed this comment: 

https://github.com/sourcegraph/zoekt/blob/acffe7943b15766abb584ebf77633d87fd283ee6/stream/stream.go#L144-L146

I copied the same pattern here, but I didn't see anything that jumped out in https://pkg.go.dev/encoding/gob for why gob encoding errors wouldn't work. However, I need to write a test that actually exercises this case.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: I ~basically~ stole bufio.Scanner's interface here. I thought it might be nice to:

1. abstract away the details about how the stream is encoded (e.g. the stream control messages for items, errors, EOF) from the caller
2. these responses could be huge, so it might be nice to allow the caller to work incrementally without having to wait for the entire response with `80k+ entries). This interface is still easy enough to work with to extract all the queueItems at once if you still want to do that (ex: the test cases)

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: I'm not too happy with these verbose names, but I think we're stuck with these unless we move them into another package.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go:142]: The current CLI  output doesn't use any fields that are unique to `*queueItem`, but my next follow up after this lands is to add a `firstInsertedIntoQueue time.Time` field to every queueItem so that it can be displayed in the output. This is why I added a new kind of iteration method to the queue struct. 

I don't think it makes sense to add this new field to IndexOptions since it's essentially metadata about the queue iteslf.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: In the future we can `switch` off of this header to perhaps return a HTML version of this page, etc.

**Comment by ggilmore** [comment]
PTAL @sourcegraph/search-core

**Comment by stefanhengl** [comment]
Demo looks super cool!

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: for the other debug commands we use fs.<Type> instead of fs.<Type>Var. Just the root flogs use the fs.<Type>Var because they set fields on the struct.

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: Maybe make the port a global const and use it here and for server?

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: I think we should use url.Parse to parse the serverAddress. Go's next release seems to come with a handy function for joining https://sourcegraph.com/github.com/golang/go/-/blob/src/net/url/url.go?L1193&subtree=true

Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/debug.go]: Yes, I think that would be nice!

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: Do we have to add another tab here?

from the tabwriter docs
>Column cells must be tab-terminated, not tab-separated: non-tab terminated trailing text at the end of a line forms a cell but that cell is not part of an aligned column

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/debug.go]: do we need localhost here or could we just ask for the port? If we just go with the port you don't have to distinguish between `serverAddress` and `fullAddress` and can just call it `addr`.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/main.go:573]: Why the rename?

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/queue.go]: Do we need to export all them? EG We don't seem to send `Indexed` to the client.

**Review by stefanhengl — COMMENTED** [review]
clicking submit before our sync all. Will continue later.

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (ggilmore) [reply cmd/zoekt-sourcegraph-indexserver/debug.go]: I thought it would be nice to have the ability to use this command on remote zoekt instances (e.g. if you were port forwarding a pod). 

What about having separate `-port` and `-hostname` flags?

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (ggilmore) [reply cmd/zoekt-sourcegraph-indexserver/main.go:573]: https://github.com/sourcegraph/zoekt/blob/73bad50487a599202231ee86851884afefd134ca/cmd/zoekt-sourcegraph-indexserver/main.go#L556-L561

I wanted the server to register both the new `/debug/queue` route (that we just added in its PR) and the original `/` route that it handled. In order to support that, I thought it would be clearer to use separate `http.HandlerFunc`s instead of single a single ServeHTTP method that switched on the route or something.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (ggilmore) [reply cmd/zoekt-sourcegraph-indexserver/queue.go]: I didn't see a good reason to leave one of these fields unexported since the type itself is unexported (we aren't leaking more information). Even though we don't use it in the CLI output for now, I can see us adding more columns in the CLI output, or returning more of this information in the html/json version of this output.

**Comment by ggilmore** [comment]
PTAL, I extracted most of the HTTP request logic into a separate function `createQueueItemStream` + switched to using httptest. I have a distinct feeling that I'm over complicating things, but maybe y'all feel differently?

**Review by keegancsmith — COMMENTED** [review]
This is really cool! I do have one bigger suggestion though to reduce the number of moving parts / expose this more.

Can you just add an HTTP endpoint which returns the text-plain output you have implemented in the debug subcmd?
Pros:
- much smaller diff
- no need for a format for communication to subprocess
- you can interact with just curl
- we can expose this information on the debug endpoint

In particular I think being able to easily tell an admin to visit one of our special proxied URLs is super powerful when debugging customer issues. The need to exec in and run a custom command makes this a bit harder to communicate.

**Comment by ggilmore** [comment]
PTAL, I ripped out the gob logic in the latest update

File: cmd/zoekt-sourcegraph-indexserver/debug.go
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/debug.go]: non-blocking design: do you imagine people actually using these flags in practice? eg to remove a headers from cli output you just do `| tail -n +2`. For connection timeout you just hit ctrl-c or just use the timeout command.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: why do you wrap with a once? There is no side effects from what I can tell, you are mutating the input mux only.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: minor: instead of writing directly to `w`, can you write to an intermediate `bytes.Buffer`. Unlikely a problem given this is an internal debug command, but holding a coarse lock while a writing to a user leads to slow users blocking up programs (or slowlaris attacks on public endpoints). I don't think we need to be concerned of memory use in this case.

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: unused code

File: cmd/zoekt-sourcegraph-indexserver/queue.go
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/queue.go]: minor suggestion for avoiding the need to comment code:
```suggestion
		// popped items go to the back
		xPopped := x.heapIdx < 0
		yPopped := y.heapIdx < 0
		if xPopped != yPopped {
			return !xPopped
		}
```

it might even read better as `xEnqueued` or `xOnQueue`, then the return is easier to understand.

**Review by keegancsmith — APPROVED** [review]
LGTM!

**Review by keegancsmith — APPROVED** [review]

**Pull request merged**
